### PR TITLE
some fixes and improvements to config reload

### PR DESCRIPTION
**Pull request by trungdtbk**
A permanent learn port fails to learn after reloading its config.

### CODE REVIEW

**Comment by gizmoguy** [comment]
Hi @trungdtbk many thanks for another patch!

Is this a fix for #496? Does it fix everything outlined there?

I assume tests are passing?

File: faucet/valve.py [hunk 1]
```diff
@@ -662,12 +662,13 @@ def _port_add_vlans(self, port, mirror_act):
...
         return ofmsgs
 
-    def port_add(self, dp_id, port_num, modify=False):
+    def port_add(self, dp_id, port_num, modify=False, old_eth_srcs=[]):
         """Handle the addition of a port.
 
         Args:
             dp_id (int): datapath ID.
             port_num (int): port number.
+            old_eth_srcs (list): eth MAC of antiproofing hosts
```
Comment (gizmoguy) [inline faucet/valve.py]: I think there's a typo here "antiproofing" -> "antispoofing"?

File: faucet/valve.py [hunk 2]
```diff
@@ -684,8 +685,14 @@ def port_add(self, dp_id, port_num, modify=False):
 
         ofmsgs = []
         if modify:
-            if not port.permanent_learn:
-                # delete eth_dst rules
+            if port.permanent_learn:
+                # delete antiproofing flow
```
Comment (gizmoguy) [inline faucet/valve.py]: similar antiproofing typo?

File: faucet/valve.py [hunk 3]
```diff
@@ -753,14 +761,19 @@ def port_delete(self, dp_id, port_num):
 
         ofmsgs = []
 
-        if not port.permanent_learn:
+        if port.permanent_learn:
+            # delete antiproofing flows
```
Comment (gizmoguy) [inline faucet/valve.py]: similar antiproofing typo?

File: faucet/valve.py [hunk 4]
```diff
@@ -1021,14 +1045,13 @@ def _get_config_changes(self, new_dp):
         for vid, new_vlan in list(new_dp.vlans.items()):
             if vid not in self.dp.vlans or new_vlan != self.dp.vlans[vid]:
                 changed_vlans.add(vid)
-            for port in new_vlan.get_ports():
-                changed_ports.add(port.number)
+                for port in new_vlan.get_ports():
+                    changed_ports.add(port.number)
```
Comment (gizmoguy) [inline faucet/valve.py]: Do you know what happened to the whitespace here? is this a correction where we had incorrect whitespace?

File: faucet/valve.py (hunk 3 above)
Comment (trungdtbk) [inline faucet/valve.py]: I will fix that

File: faucet/valve.py (hunk 4 above)
Comment (trungdtbk) [inline faucet/valve.py]: This was a bug. Without the whitespace, we reload all ports no matter what their change status

**Comment by trungdtbk** [comment]
Hi Brad,
I fixed the typos. Yes, it passes all tests and it is for the issue #496

**Comment by gizmoguy** [comment]
Beautiful, thanks again @trungdtbk

**Comment by anarkiwi** [comment]
Something got broken here:

00:25:09     @josh | Traceback (most recent call last):

                                   │
00:25:09     @josh |   File "/home/faucet/faucet/faucet/valve_util.py",
line 52, in __koe
                                       │
00:25:09     @josh |     func(*args, **kwargs)

                                    │
00:25:09     @josh |   File "/home/faucet/faucet/faucet/faucet.py", line
494, in handler_connect_or_disconnect
                                      │
00:25:09     @josh |     self.handler_datapath(ryu_dp)

                                    │
00:25:09     @josh |   File "/home/faucet/faucet/faucet/faucet.py", line
519, in handler_datapath
                                       │
00:25:09     @josh |     dp_id, discovered_up_port_nums)

                                    │
00:25:09     @josh |   File "/home/faucet/faucet/faucet/valve.py", line
575, in datapath_connect
                                        │
00:25:09     @josh |     ofmsgs.extend(self._apply_config_changes(self.dp,
changes))
                                    │
00:25:09     @josh |   File "/home/faucet/faucet/faucet/valve.py", line
1096, in _apply_config_changes
                                        │
00:25:09     @josh |     old_eth_srcs =
self._get_eth_srcs_learned_on_port(old_dp, port_no)
                                                                       │
00:25:09     @josh |   File "/home/faucet/faucet/faucet/valve.py", line
1000, in _get_eth_srcs_learned_on_port
                                        │
00:25:09     @josh |     vlan = dp.vlans[vid]

                                   │
00:25:09     @josh | KeyError: None

Please investigate.

**Comment by anarkiwi** [comment]
Tests do not pass.


EEEEEEEEEEEEEEEEEEEEEEEEE
======================================================================
ERROR: test_vlan_acl_deny (__main__.ValveACLTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_invalid_vlan (__main__.ValveReloadConfigTestCase)
Test that packets with incorrect vlan tagging get dropped.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_known_eth_dst_rule (__main__.ValveReloadConfigTestCase)
Test that packets with known eth dst addrs are output correctly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_known_eth_dst_rule_deletion (__main__.ValveReloadConfigTestCase)
Test that eth dst rules are deleted when the mac is learned on
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_known_eth_src_deletion (__main__.ValveReloadConfigTestCase)
Verify that when a mac changes port the old rules get deleted.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_known_eth_src_rule (__main__.ValveReloadConfigTestCase)
test that packets with known eth src addrs are not sent to controller.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_mac_learning_vlan_separation
(__main__.ValveReloadConfigTestCase)
Test that when a mac is seen on a second vlan the original vlan
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_port_acl_deny (__main__.ValveReloadConfigTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_port_add_input (__main__.ValveReloadConfigTestCase)
test that when a port is enabled packets are input correctly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_port_delete_eth_dst_removal (__main__.ValveReloadConfigTestCase)
Test that when a port is disabled packets are correctly output.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_port_down_eth_src_removal (__main__.ValveReloadConfigTestCase)
Test that when a port goes down and comes back up learnt mac
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_unknown_eth_dst_rule (__main__.ValveReloadConfigTestCase)
Test that packets with unkown eth dst addrs get flooded correctly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_unknown_eth_src (__main__.ValveReloadConfigTestCase)
Test that packets from unknown macs are sent to controller.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 668, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: None

======================================================================
ERROR: test_invalid_vlan (__main__.ValveTestCase)
Test that packets with incorrect vlan tagging get dropped.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_known_eth_dst_rule (__main__.ValveTestCase)
Test that packets with known eth dst addrs are output correctly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_known_eth_dst_rule_deletion (__main__.ValveTestCase)
Test that eth dst rules are deleted when the mac is learned on
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_known_eth_src_deletion (__main__.ValveTestCase)
Verify that when a mac changes port the old rules get deleted.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_known_eth_src_rule (__main__.ValveTestCase)
test that packets with known eth src addrs are not sent to controller.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_mac_learning_vlan_separation (__main__.ValveTestCase)
Test that when a mac is seen on a second vlan the original vlan
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_port_acl_deny (__main__.ValveTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_port_add_input (__main__.ValveTestCase)
test that when a port is enabled packets are input correctly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_port_delete_eth_dst_removal (__main__.ValveTestCase)
Test that when a port is disabled packets are correctly output.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_port_down_eth_src_removal (__main__.ValveTestCase)
Test that when a port goes down and comes back up learnt mac
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_unknown_eth_dst_rule (__main__.ValveTestCase)
Test that packets with unkown eth dst addrs get flooded correctly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

======================================================================
ERROR: test_unknown_eth_src (__main__.ValveTestCase)
Test that packets from unknown macs are sent to controller.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test_valve.py", line 123, in setUp
    range(1, self.NUM_PORTS + 1)
  File "/home/faucet/faucet/faucet/valve.py", line 575, in datapath_connect
    ofmsgs.extend(self._apply_config_changes(self.dp, changes))
  File "/home/faucet/faucet/faucet/valve.py", line 1096, in
_apply_config_changes
    old_eth_srcs = self._get_eth_srcs_learned_on_port(old_dp, port_no)
  File "/home/faucet/faucet/faucet/valve.py", line 1000, in
_get_eth_srcs_learned_on_port
    vlan = dp.vlans[vid]
KeyError: 'v100'

----------------------------------------------------------------------
Ran 25 tests in 0.380s

**Comment by anarkiwi** [comment]
I reverted the changes, please do rebase and submit again when fixed.

**Comment by trungdtbk** [comment]
Hi Josh,
There is something inconsistent that causes it fails the "test_valve.py". The port config refers to a vlan identifier which is not used as the key for vlans.
vlans:
    v100:
         vid: 0x100
But...
    port_1:
          native_vlan: v100
As we load config, "v100" is not used as the key for vlan dict. I suggest that we should stick to one format only, either the previous one or:
vlans:
     100:
          name: "abc"
          description: "abcde"
port_1:
     native_vlan: 100

**Comment by KitL** [comment]
How it is represented internally shouldnt necessarily affect how it is
configured.

The reason it is this way is for backwards compatibility. I wanted to move
to using names to represent vlans and ports rather than numbers, but I
didnt want to break every existing currently being used in production.

Is this really something you cant work around?

**Comment by anarkiwi** [comment]
Agreed with @KitL. This change broke my production setup, so whatever the
problem is, it affects more than the test. We need to get to the root cause.

**Comment by trungdtbk** [comment]
I mean when we read the config, we add "v100" as the reference to vlan in Port object. But "v100" doesn't exist anywhere. We use vid as the key in vlans dict. I know that should not matter. But we need to fix the config parser for consistency

**Comment by anarkiwi** [comment]
If you have found a bug in the parser (where it crashes if something is
inconsistent), then we would come a fix that causes that inconsistency to
be logged (eg, "port xyz undefined").

**Pull request merged**